File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        dic = {')': '(', ']': '[', '}': '{'}    
        for i in range(len(s)):
            if not dic.get(s[i]):
                stack.append(s[i])
            elif not stack or dic.get(s[i]) != stack[-1]:
                return False
            elif stack and dic.get(s[i]) == stack[-1]:
                stack.pop()
        return True if not stack else False

File: 0020-valid-parentheses/0020-valid-parentheses.kt
import java.util.*
import java.util.Deque
import java.util.ArrayDeque

class Solution {
    fun isValid(s: String): Boolean {
        val stack: Deque<Char> = ArrayDeque()
        val table: Map<Char, Char> = mapOf(
            ')' to '(',
            '}' to '{',
            ']' to '['
        )

        for (i in s.indices) {
            if (!table.containsKey(s[i])) {
                stack.push(s[i])
            } else if (stack.isEmpty() || table[s[i]] != stack.pop()) {
                return false
            }
        }
        return stack.isEmpty()
    }
}
